commit b5bbac1a9b07016602559ff483df265fef6c1f83
Author: Ionut Nicu <ioan.nicu.ext@nsn.com>
Date:   Mon Jan 13 12:00:08 2014 +0100

    ext4fs: fix "invalid extent block" error
    
    For files where we actually have extent indexes following
    an extent header (ext_block->eh_depth != 0), the do/while
    loop from ext4fs_get_extent_block() does not select the
    proper extent index structure.
    
    For example, if we have:
    
    ext_block->eh_depth = 1
    ext_block->eh_entries = 1
    fileblock = 0
    index[0].ei_block = 0
    
    the do/while loop will exit with i set to 0 and the
    ext4fs_get_extent_block() function will return 0, even if
    there was a valid extent index structure following the
    header.
    
    Signed-off-by: Ionut Nicu <ioan.nicu.ext@nsn.com>
    Signed-off-by: Mathias Rulf <mathias.rulf@nsn.com>

diff --git a/fs/ext4/ext4_common.c b/fs/ext4/ext4_common.c
index c5e654235a..02da75c084 100644
--- a/fs/ext4/ext4_common.c
+++ b/fs/ext4/ext4_common.c
@@ -1430,7 +1430,7 @@ static struct ext4_extent_header *ext4fs_get_extent_block
 			i++;
 			if (i >= le16_to_cpu(ext_block->eh_entries))
 				break;
-		} while (fileblock > le32_to_cpu(index[i].ei_block));
+		} while (fileblock >= le32_to_cpu(index[i].ei_block));
 
 		if (--i < 0)
 			return 0;
